pkg/common/worker.go

```diff
@@ -0,0 +1,91 @@
+package common
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+// Errors that may occur when sending tasks to a worker.
+var (
+	ErrWorkerClosed  = errors.New("worker is closed")
+	ErrWorkerTooBusy = errors.New("worker is already overloaded")
+)
+
+// Configuration for the worker.
+type WorkerConfig[T any] struct {
+	// The size of the bounded channel.
+	ChannelSize int
+	// Timeout after which `OnTimeout` is called.
+	Timeout time.Duration
+	// A closure that is called once `Timeout` is reached.
+	OnTimeout func()
+	// A closure that is executed upon reception of a task.
+	OnTask func(T)
+}
+
+// We need to wrap the channel in a struct so that we can close it from the outside and
+// check by the sender if the channel is closed (there is no elegant way to do it in Go).
+type Worker[T any] struct {
+	channel chan<- T
+	mutex   sync.Mutex
+	closed  bool
+}
+
+// Stop the channel unless already closed.
+func (c *Worker[T]) Stop() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if !c.closed {
+		close(c.channel)
+		c.closed = true
+	}
+}
+
+// Send a task to the worker. Returns `true` if the task
+// has been sent, `false` if the channel is already closed.
+func (c *Worker[T]) Send(task T) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// If the channel is not closed, check if we can send the task.
+	if !c.closed {
+		// We don't want to block here since it's the whole point of this
+		// component (that the CPU bound tasks are handled by the worker).
+		select {
+		case c.channel <- task:
+			return nil
+		default:
+			return ErrWorkerTooBusy
```

Comment:
Will this risk dropping the most recent packets and therefore the SFU ending up permanently behind? We presumably still need to keep this a last resort, but probably need the worker OnTask function to drop the oldest packets in the queue in order to catch up?

Reply:
Yes, you're right, the approach that you suggest is also something that I think is a better choice. The only reason I did not implement it is that I thought that it might be a bit complicated (especially with packet rewriting), so it probably requires a bit more thinking, so I thought a "smarter" way (the way you suggested) to drop older packets is better to implement once we land the support of keyframe detection and switching on key frames since it would allow us to implement the logic (drop all non-key-frame packets from the queue when the queue gets full too quickly).

One thing to note is that if we really return `ErrWorkerTooBusy`, then it would mean that the conference is quite overloaded already and the SFU can't keep up with the new subscriptions. So ideally it should not happen at all for conferences, I may assume that we may get this error only when there are too many things going on in the SFU.